File: lib/java/DNJavaConverter.js
import { InputStream, CommonTokenStream, tree as _tree } from 'antlr4'
import Java9Lexer from '../../parser/java/Java9Lexer.js'
import Java9Parser from '../../parser/java/Java9Parser.js'
import DNJavaParserListener from './DNJavaParserListener.js'
import { parentPort, workerData } from 'worker_threads'

export default class DNJavaConverter {
    convert(content, cb) {
        if (!content) {
            if (!workerData) {
                if (cb) {
                    cb(null, null)
                }
                return
            }
            content = workerData.content
        }
        if (!cb) {
            cb = this.callback
        }

        try {
            const chars = new InputStream(content)
            const lexer = new Java9Lexer(chars)

            const tokens = new CommonTokenStream(lexer)
            const parser = new Java9Parser(tokens)
            const tree = parser.compilationUnit()
            let isPartOfFile = workerData ? workerData.isPartOfFile : false
            const listener = new DNJavaParserListener(cb, content)
            _tree.ParseTreeWalker.DEFAULT.walk(listener, tree)
        } catch (e) {
            cb(null, e)
        }
    }

    callback(result, error) {
        // Send a message to the main thread.
        parentPort.postMessage({ result: result, error: error });
    }
}

new DNJavaConverter().convert()

File: lib/src/java/dn_java_converter.dart
import 'dart:async';
import 'dart:io';

import 'package:antlr4/antlr4.dart';
import 'package:dart_native_codegen/parser/java/Java9Lexer.dart';
import 'package:dart_native_codegen/parser/java/Java9Parser.dart';

import '../common.dart';
import 'DartJavaCompiler.dart';
import 'dn_java9_parser_listener.dart';

class DNJavaConverter {
  static Future<String> convert(JavaFile file) async {
    file.fileType = FILE_TYPE.source_file;
    String content = File(file.path).readAsStringSync();
    Completer<String> completer = Completer();
    try {
      final chars = InputStream.fromString(content);
      final lexer = Java9Lexer(chars);
      final tokens = CommonTokenStream(lexer);
      final parser = Java9Parser(tokens);
      parser.addErrorListener(DiagnosticErrorListener());
      parser.buildParseTree = true;
      final tree = parser.compilationUnit();
      Callback cb = (String content, {String error}) {
        if (content != null) {
          completer.complete(content);
        } else {
          completer.completeError(error, StackTrace.current);
        }
      };
      ParseTreeWalker.DEFAULT.walk(DNJavaParserListener(file, cb), tree);
    } catch (e) {
      completer.completeError(e, StackTrace.current);
    }
    return completer.future;
  }
}